### DISCUSSION TITLE
Best way to use sparse matrix as counter

### DISCUSSION THREAD

**Discussion by mtseng15** [vbarrielle/sprs#352] [General]
Hello all,

First of all, thank you very much for writing this crate. It's awesome. I am still learning Rust (and how to use this crate) and I had a question about the best way to do something. 

I am working on performing some information theory calculations where I must increment entries in a 2D sparse matrix. Right now I have a function (see below) that uses `get_mut()`, but there are notes in the docs that it is not very efficient and doing a little profiling shows this is a little slow. Is there a better way to do this? 

Thanks a lot!

```rust

// Instantiation of data
let largest_word = u64::pow(2, 16 as u32) as usize;
let data:CsMat<u32> = CsMat::zero((largest_word, largest_word)).to_csc();

// ...


// Function to perform incrementation
fn increment(&mut data, row: &usize, col: &usize, n: u32) {
    match data.get_mut(*row,*col) {
        None => {
            data.insert(*row,*col,n);
        },
        Some(val) => {
            *val += n;
        }
    }
}
```

**Comment by mulimoen** [comment]
This is slow since the indexing operation can be expensive. It might be faster to use a hashmap and create a sparse matrix from that. You need to benchmark, it depends on your problem if it is faster or not

**Reply by mtseng15** [reply]
@mulimoen Thanks for your response! 

I was originally (still am) using a hashmap and I'm benching using sprs against it. The hashmap is faster, but far more memory consumptive; though I guess that is kinda the idea behind a hashmap: trading space for speed. 

I wanted to make sure I was using your crate optimally to give it the best comparison.  

Do you have any general tips for improving performance with sprs?

**Reply by mulimoen** [reply]
The less lookups you do the better, if you can index along an outer dimension all at once that might be faster than doing random read/writes.

**Reply by mtseng15** [reply]
Thanks for your response! Sadly the application is literally counting random read/writes, lol. But that is good to know.
